### PR TITLE
Signaling

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
This implements the signaling/human-in-the-loop feature of Flyte on the flytekit side.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Ported over the new flyteidl messages and created a new GateNode class.

Can add a gating element in a few different ways

### Run time user input
```python
from flytekit.core.gate import sleep, wait_for_input, approve

@workflow
def wf():
    s1 = wait_for_input("my-signal-name", timeout=timedelta(hours=1), expected_type=bool)
    y = t2(some_bool=s2)
```
### Sleep
```python
@workflow
def wf_sleep():
    x = sleep(timedelta(seconds=10))
    b = t1(a=5)
    x >> b
```

### Approve prior task output
```python
@workflow
def wf(a: int) -> int:
    x = t1(a=a)
    q = t2(a=approve(x, "approval name", timeout=timedelta(hours=2)))
    return q
```

#### Nit
Updated the github workflow to allow plugins to pick up the new idl.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/208

## Follow ups
https://github.com/flyteorg/flyte/issues/3124 (Refactor out the pyflyte run string parsing code)
https://github.com/flyteorg/flyte/issues/3125 (FlyteRemote signaling to fulfill/approve signals of running workflows)

### CODE REVIEW

File: flytekit/core/gate.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    # This is to satisfy the LocallyExecutable protocol
+    def local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], Promise, VoidPromise]:
+        if self.sleep_duration:
+            print(f"Fake sleeping {self.name}")
+            return VoidPromise(self.name)
+
+        # Trigger stdin
+        if issubclass(self.input_type, bool):
+            input(f"Pausing execution for gate {self.name}, press enter to continue...\n")
```
Comment (kumare3) [inline flytekit/core/gate.py:84]: click.prompt

File: tests/flytekit/unit/core/test_gate.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        return a + 5
+
+    @task
+    def t2(a: int) -> int:
+        return a + 6
+
+    @workflow
+    def wf(a: int) -> typing.Tuple[int, int]:
+        x = t1(a=a)
+        s1 = signal("my-signal-name", timeout=timedelta(hours=1), expected_type=bool)
```
Comment (kumare3) [inline tests/flytekit/unit/core/test_gate.py:37]:
```suggestion
        s1 = flytekit.wait_for("my-signal-name", timeout=timedelta(hours=1), expected_type=bool)
```

File: flytekit/core/gate.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+from flytekit.models.core import workflow as _workflow_model
+from flytekit.models.types import LiteralType
+
+DEFAULT_TIMEOUT = datetime.timedelta(hours=1)
+
+
+class Gate(object):
+    """
+    A gate needs to behave kind of like a task. Instead of running code however, it either needs to wait for user
+    input, or run a timer.
```
Comment (cosmicBboy) [inline flytekit/core/gate.py:21]:
```suggestion
    A node type that waits for user input before proceeding with a workflow.

    A gate is a type of node that behaves like a task, but instead of running code, it either needs to wait for user input to proceed or wait for a timer to complete running.
```

File: flytekit/core/gate.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    Create a Gate object. This object will function like a task. Note that unlike a task,
+    each time this function is called, a new Python object is created. If a workflow
+    calls a subworkflow twice, and the subworkflow has a signal, then two Gate
+    objects are created. This shouldn't be a problem as long as the objects are identical.
+
+    :param name:
+    :param timeout:
+    :param expected_type:
+    :return:
```
Comment (cosmicBboy) [inline flytekit/core/gate.py:118]: missing parameter descriptions

**Review by cosmicBboy — COMMENTED** [review]
Adding overall feedback on the flytekit signaling syntax: tbh it's fairly unintuitive to me, and I had to read it several times (+ get a walkthrough from Yee) to fully grok it.

From my understanding, there are 2 types of ways gate nodes are used:

1. to wait for user input, which is explicitly passed onto the downstream task
2. to wait for user acceptance, which unblocks the flow of inputs from one task to another

Just spit-balling here, but I'd find this syntax a lot more readable:

```python
from flytekit import task, workflow, wait_on

@task
def t1(a: int) -> int: return a + 5

@task
def t2(a: int) -> int: return a + 6

@workflow
def wf(a: int) -> typing.Tuple[int, int]:
    x = t1(a=a)
    y = t2(a=wait_for_input(int, name="user_input_name", timeout=timedelta(hours=2)))
    z = t1(a=wait_for_approval(x, "approval_name", timeout=timedelta(hours=1)))
    return y, z


# versus
@workflow
def wf(a: int) -> typing.Tuple[int, int]:
     x = t1(a=a)
     s1 = signal("my-signal-name", timeout=timedelta(hours=1), expected_type=bool)
     s2 = signal("my-signal-name-2", timeout=timedelta(hours=2), expected_type=int)
     z = t1(a=5)
     y = t2(a=s2)
     x >> s1
     s1 >> z
     return y, z
```

A few things to note about this syntax:
- the first arg to `wait_for_input` is the expected type the user needs to provide
- the first arg to `wait_for_approval` is the promise that is being blocked by the gate
- The two functions above clarifies to the user what exactly the workflow is waiting on. NOTE: they could be reduced to a single function `wait_for(promise_or_type, ...)` or `wait_for(promise=None, user_input_type=None)` where only one argument cat be provided.
- the syntax allows for in-line gate nodes that don't need to be strung together with `>>` (this introduces additionally cognitive burden imho, where one need to keep track of nodes being defined _and_ how they're composed together outside of regular Python function call syntax.)

**Comment by wild-endeavor** [comment]
cc @hamersaw 
this is the final approach we're going for.
```python
from flytekit import task, workflow, approve, wait_for_input

@task
def t1(a: int) -> int: return a + 5

@task
def t2(a: int) -> int: return a + 6

# proposal
@workflow
def wf(a: int) -> typing.Tuple[int, int]:
    x = t2(a=wait_for_input(int, name="user_input_name", timeout=timedelta(hours=2)))
    
    # by default dan forgot to update the comments on this section
    y = t1(a=a)
    z = t1(a=approve(y, name="approval_name", timeout=timedelta(hours=1)))

    return x, z
```

**Review by eapolinario — APPROVED** [review]
